driver/src/main/java/org/neo4j/driver/internal/connector/socket/SocketClient.java

```diff
@@ -146,7 +140,9 @@ private SocketProtocol negotiateProtocol() throws IOException
 
         while ( buf.remaining() > 0 )
         {
+            int start = buf.position();
             channel.read( buf );
+            logger.trace( "S: " + BytePrinter.hex( buf, start, buf.position() ) );
```

Comment:
We can't do these like this - this will print every bit of the buffer even if `trace` is off. 

I'd suggest we do this the same way you did it with the Input/Output streams - introduce classes that implement WritableByteChannel/ReadableByteChannel, and can wrap the regular channels. Then, when we create the real channels, we do something like:

    if( logger.isTraceEnabled() )
    {
        input = new LoggingInputChannel( logger, input );
        output = new LoggingOutputChannel( logger, output );
    }

That way, we don't have to intermingle these log statements with the rest of the code, and there's zero overhead if trace logging is not enabled.